### PR TITLE
sql: don't copy old pk when changing bucket size on sharded index

### LINKED ISSUE #45889: sql: don't create old pk as an unique index when alter primary key of hash sharded index
Consider this table:
```SQL
CREATE TABLE events6 (
             |     product_id INT8 NOT NULL,
             |     owner UUID NOT NULL,
             |     serial_number VARCHAR NOT NULL,
             |     event_id UUID NULL,
             |     ts TIMESTAMP NULL,
             |     data JSONB NULL,
             |     CONSTRAINT "primary" PRIMARY KEY (product_id ASC, owner ASC, serial_number ASC) USING HASH WITH BUCKET_COUNT = 8,
             |     FAMILY "primary" (product_id, owner, serial_number, event_id, ts, data, crdb_internal_owner_product_id_serial_number_shard_8)
             | )
```
Let's say we got the bucket size wrong and want to adjust it to 7:
```SQL
ALTER TABLE events6 ALTER PRIMARY KEY USING COLUMNS (product_id ASC, owner ASC, serial_number ASC) USING HASH WITH BUCKET_COUNT = 7;
NOTICE: primary key changes spawn async cleanup jobs. Future schema changes on "events6" may be delayed as these jobs finish
ALTER TABLE

Time: 147.706ms
```
Now, if we look at the table:
```SQL
CREATE TABLE events6 (
             |     product_id INT8 NOT NULL,
             |     owner UUID NOT NULL,
             |     serial_number VARCHAR NOT NULL,
             |     event_id UUID NULL,
             |     ts TIMESTAMP NULL,
             |     data JSONB NULL,
             |     CONSTRAINT "primary" PRIMARY KEY (product_id ASC, owner ASC, serial_number ASC) USING HASH WITH BUCKET_COUNT = 7,
             |     UNIQUE INDEX events6_crdb_internal_owner_product_id_serial_number_shard_8_product_id_owner_serial_number_key (product_id ASC, owner ASC, serial_number ASC) USING HASH WITH BUCKET_COUNT = 8,
             |     FAMILY "primary" (product_id, owner, serial_number, event_id, ts, data, crdb_internal_owner_product_id_serial_number_shard_8, crdb_internal_owner_product_id_serial_number_shard_7)
             | )
```
We see a new unique index (the old primary key) with the bucket count =8 and not 7. In the non-hash sharded index implementation we keep this index around to make sure we don't regress performance. In the hash sharded world, you are changing this to change the bucket count for performance. Like interleaving, I think we should not build this index by default.

Version: V20.1.0-BETA.2-605-G01F131D018

Also considered:
We could also change the new unique index to the new bucket count but I think its an unnecessary index.

### PR DESCRIPTION
**Pull request by rohany**
Fixes #45889.

Release note (sql change): This PR causes primary key changes to
not create a copy of the old primary key if the primary key change
only changes the bucket count of a hash sharded index.

### CODE REVIEW

**Comment by otan** [comment]
Just for sanity - why do we not want this behaviour?

**Comment by rohany** [comment]
The argument is as follows:

We make a copy of the old primary key so that your query performance doesn't regress after the primary key change. However, in the hash sharded case, the change being made is going to affect performance (i.e. what is the bucket count, or what columns are hashed). Keeping around the old primary key means keeping around the old primary key with the incorrect number of buckets etc. What does @ajwerner think about this?

**Review by ajwerner — COMMENTED** [review]
I agree with @rohany here, at least for the tested case. I can't see any reason why a user would want to retain the old bucket count.

Is there are case where the user would want to retain the index if they changed the set of columns but both were hashed? That one is less clear to me.


Consider:

```
CREATE TABLE t (x INT NOT NULL, y INT NOT NULL, ts TIMESTAMPTZ NOT NULL, PRIMARY KEY (ts, x, y) USING HASH WITH BUCKET_COUNT = 2);
```
then:

```
ALTER TABLE t ALTER PRIMARY KEY USING COLUMNS (ts, y, x) USING HASH WITH BUCKET_COUNT=3
```

I could see it either way in this case. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45894#-:-M2-fszm9i-iuJ_-Iy0I:b-7pn27k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

File: pkg/sql/alter_table.go [hunk 1]
```diff
@@ -563,7 +563,9 @@ func (n *alterTableNode) startExec(params runParams) error {
 			// does, but doesn't store anything. We only recreate the index if
 			// the table has a primary key (no DROP PRIMARY KEY statements have
 			// been executed), and the primary key is not the default rowid key.
```
Comment (otan) [inline pkg/sql/alter_table.go]: nit: maybe updated comment for hash sharded cases too

**Review by otan — APPROVED** [review]

**Comment by jordanlewis** [comment]
Does this want to be merged?

**Comment by rohany** [comment]
Reviving this PR! I am more satisfied with it now -- I changed it to only not create a copy if only the bucket count of the hash sharded index is changing. PTAL

File: pkg/sql/alter_primary_key.go [hunk 2]
```diff
@@ -327,3 +325,20 @@ func (p *planner) AlterPrimaryKey(
...
+// * The table has a primary key (no DROP PRIMARY KEY statements have
+//   been executed).
+// * The primary key is not the default rowid primary key.
+// * The new primary key isn't the same hash sharded primary key with a
+//   different bucket count.
+func shouldCopyPrimaryKey(desc *MutableTableDescriptor, newPK *sqlbase.IndexDescriptor) bool {
+	oldPK := desc.PrimaryIndex
+	return desc.HasPrimaryKey() && !desc.IsPrimaryIndexDefaultRowID() &&
+		!(oldPK.IsSharded() && newPK.IsSharded() &&
+			// The first column in the columnIDs is the shard column, which will be different.
```
Comment (otan) [inline pkg/sql/alter_primary_key.go:340]: this was a teensy bit hard to read but makes sense after squinting. 

wondering if breaking down to the `if` statements might be better... but it's personal taste for me.

````

func shouldCopyPrimaryKey(desc *MutableTableDescriptor, newPK *sqlbase.IndexDescriptor) bool {
  // no primary key means we have nothing to copy
  if !desc.HasPrimaryKey() {
     return false
  }
  // we never want to copy default row id
  if desc.IsPrimaryIndexDefaultRowID() {
    return false
  }
  if oldPK.IsSharded() && newPK.IsSharded() {
     // ...
  }
  return true
}
```

**Review by otan — COMMENTED** [review]
LGTM. Completely optional bikeshedding comment.

File: pkg/sql/alter_primary_key.go (hunk 2 above)
Comment (rohany) [reply pkg/sql/alter_primary_key.go:340]: I'm down for this, I find it hard to read as well.

**Comment by rohany** [comment]
bors r=otan

**Pull request merged**